Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -141,76 +152,77 @@ std::string DecodeString(pb_istream_t* stream) {
   return result;
 }
 
-}  // namespace
-
-using firebase::firestore::model::FieldValue;
-
-void Serializer::EncodeFieldValue(const FieldValue& field_value,
-                                  std::vector<uint8_t>* out_bytes) {
-  // TODO(rsgowman): how large should the output buffer be? Do some
-  // investigation to see if we can get nanopb to tell us how much space it's
-  // going to need.
-  uint8_t buf[1024];
-  pb_ostream_t stream = pb_ostream_from_buffer(buf, sizeof(buf));
-
+// Named '..Impl' so as to not conflict with Serializer::EncodeFieldValue.
+// TODO(rsgowman): Refactor to use a helper class that wraps the stream struct.
+// This will help with error handling, and should eliminate the issue of two
+// 'EncodeFieldValue' methods.
+void EncodeFieldValueImpl(pb_ostream_t* stream, const FieldValue& field_value) {
   // TODO(rsgowman): some refactoring is in order... but will wait until after a
   // non-varint, non-fixed-size (i.e. string) type is present before doing so.
   bool status = false;
   switch (field_value.type()) {
     case FieldValue::Type::Null:
-      status = pb_encode_tag(&stream, PB_WT_VARINT,
+      status = pb_encode_tag(stream, PB_WT_VARINT,
                              google_firestore_v1beta1_Value_null_value_tag);
       if (!status) {
         // TODO(rsgowman): figure out error handling
         abort();
       }
-      EncodeNull(&stream);
+      EncodeNull(stream);
       break;
 
     case FieldValue::Type::Boolean:
-      status = pb_encode_tag(&stream, PB_WT_VARINT,
+      status = pb_encode_tag(stream, PB_WT_VARINT,
                              google_firestore_v1beta1_Value_boolean_value_tag);
       if (!status) {
         // TODO(rsgowman): figure out error handling
         abort();
       }
-      EncodeBool(&stream, field_value.boolean_value());
+      EncodeBool(stream, field_value.boolean_value());
       break;
 
     case FieldValue::Type::Integer:
-      status = pb_encode_tag(&stream, PB_WT_VARINT,
+      status = pb_encode_tag(stream, PB_WT_VARINT,
                              google_firestore_v1beta1_Value_integer_value_tag);
       if (!status) {
         // TODO(rsgowman): figure out error handling
         abort();
       }
-      EncodeInteger(&stream, field_value.integer_value());
+      EncodeInteger(stream, field_value.integer_value());
       break;
 
     case FieldValue::Type::String:
-      status = pb_encode_tag(&stream, PB_WT_STRING,
+      status = pb_encode_tag(stream, PB_WT_STRING,
                              google_firestore_v1beta1_Value_string_value_tag);
       if (!status) {
         // TODO(rsgowman): figure out error handling
         abort();
       }
-      EncodeString(&stream, field_value.string_value());
+      EncodeString(stream, field_value.string_value());
+      break;
+
+    case FieldValue::Type::Object:
+      // NB: submessages use a wiretype of PB_WT_STRING
```

Comment:
I'll get rid of it. I added it for myself, as I found it mildly surprising that the wiretype for strings and submessages was the same. (Unlike the fact that a varint is used for bools, since I expect bools to be represented by ints.) But now that I've played around with nanopb a bit more, it seems less surprising now then it did before, since both strings and submessages are lengths followed by some arbitrary data.